### ISSUE TITLE
Тонкий клиент

### ISSUE THREAD

**Issue by aurusov** [LeKaitoW/raox#154]
Хочется научиться запускать модель через браузер. Как минимум
- отправить с клиента модель (или выбрать из фиксированного списка)
- пока идет прогон, наблюдать за статусом моделирования (работает/завершено)
- после прогона получить результаты

Следующий левел, с понижением приоритета
- пока идет прогон, менять режим моделирования и скорость
- добавить анимацию на клиенте
- добавить просмотр графиков
- добавить редактор модели и др. фичи полноценной IDE через браузер

**Comment by aurusov** [comment]
@k-alexandrovsky 
Помнишь, ты показывал работу с эклипсом через браузер ? Но там нужна была старая версия swt или еще какой-то либы. Подскажи где посмотреть и есть ли возможность запустить текущее РДО на этой базе ?

**Comment by aurusov** [comment]
Кажется, что надо встраивать сервер прямо в РДО. Это позволит во время прогона по сокету отдавать анимацию, трассировку и графики. В Эклипсе для этого есть [инструментарий](http://www.eclipse.org/webtools/).

**Comment by k-alexandrovsky** [comment]
Мне кажется, что интерфейс придётся делать самим, запускать в качестве сервера [jetty](http://www.eclipse.org/jetty/), а там уже любой фреймворк, либо spring, либо gwt. Анимация вроде более-менее абстрагирована, так что html5 canvas нарисует не хуже swt. Естественно, все будет на ajax'e. А редактор можно встроить их родной - http://www.eclipse.org/Xtext/documentation/330_web_support.html

**Comment by aurusov** [comment]
Не понял про spring/gwt.

**Comment by k-alexandrovsky** [comment]
Ну насколько я понимаю, в любом случае придется писать фронтэнд. На любой из технологий. Можно хоть на чистом JSP.

Эклипс вроде как не дает полноценную IDE в браузере. У них было это в качестве эксперимента несколько лет назад, но загнулось (видимо, за ненадобностью). Из современного я нашел вот такое - https://www.eclipse.org/che/technology/, но оно скорее всего еще очень сырое, плюс икстекст с ним вряд ли сам подружится.

Мне самый безболезненный вариант видится таким - в эклипсе на удаленном компьютере запускается сервер (составляющий вместе с эклипсом бэкэнд), который управляет прогоном, и отдает актуальную информацию на фронтэнд. Там уже рисуются джаваскриптом графики, им же анимация (`SWTAnimationContext` превращается в какой-нибудь `JavascriptCallbackAnimationContext`). А в качестве редактора моделей прикручивается Orion. Это не будет полноценная IDE для всего на свете, какой является эклипс, но возможности РДО ей покрыть можно, мне кажется.

**Comment by aurusov** [comment]
> Мне самый безболезненный вариант видится таким - в эклипсе на удаленном компьютере запускается сервер (составляющий вместе с эклипсом бэкэнд), который управляет прогоном, и отдает актуальную информацию на фронтэнд.

Тут понятно, согласен

> Там уже рисуются джаваскриптом графики, им же анимация (SWTAnimationContext превращается в какой-нибудь JavascriptCallbackAnimationContext).

Что значит там же ? Понятно, что надо попросить анимацию генерить команды отрисоки для клиента. Но рисоваться они должны же на клиенте. Что-то не понимаю твою мысль.

> А в качестве редактора моделей прикручивается Orion. Это не будет полноценная IDE для всего на свете, какой является эклипс, но возможности РДО ей покрыть можно, мне кажется.

Редактор не нужен. Пока просто запуск + состояние прогона + анимация.

**Comment by k-alexandrovsky** [comment]
> Что значит там же ? Понятно, что надо попросить анимацию генерить команды отрисоки для клиента. Но рисоваться они должны же на клиенте. Что-то не понимаю твою мысль.

Я не знаю, как на текущий момент устроены результаты с точки зрения архитектуры, но с анимацией можно сделать следующим образом -- какой-нибудь `BackendAnimationContext` вместо отрисовки фрейма переводит его в сериализованный вид (JSON/BSON/что-угодно) и начинает его хранить, чтобы потом отдать на клиент. Клиент, соответственно, рисует. Можно делать не полный фрейм каждый раз, а вычислять диффы на сервере и отдавать их. Если сервер мощный, то замедления не будет, а применять дифф дешево.

То есть идея такая, нужен еще один уровень абстракции:

```
RAO Frame (xbase) => Сцена (примитивы) => Отрисовка или сериализация (т.е. текущие контексты)
```

**Comment by aurusov** [comment]
Да, именно так и я себе представляю. Когда ты начал говорить непонятные слова, думал у тебя есть готовый стек технологий для этого. Велосипедить не хочется.

**Comment by aurusov** [comment]
1. На базе [Jetty](http://www.eclipse.org/jetty//) поднять сервер, не связывая его с raox. Настроить [nginx](https://github.com/aurusov/raox/wiki/nginx) в качестве прокси. Разработать ручку `ping`, добиться ответа. [Пример](https://github.com/aurusov/poco-http-server) простейшего сервера.
2. Добавить ручку `currentTime`, которая будет возвращать константу 12345.
3. Связать сервер с raox, чтобы сервер возвращал текущее модельное время или фразу `сервер недоступен`.

**Comment by aurusov** [comment]
@bogachev-pa 
Паша, где в коде место, которое отрабатывает один после загрузки Эклипса с РДО ? Нужно дописать код в момент запуска и выгрузки Эклипса.

**Comment by bogachev-pa** [comment]
@aurusov

https://github.com/aurusov/raox/blob/release/current/ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/RaoActivatorExtension.java#L45

**Comment by aurusov** [comment]
Добавить javasript, который сам ходит за `/currentTime` и выводит на экран через html.
